keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -957,55 +961,84 @@ complete their service.
+
+Whenever a kube-apiserver starts up, it checks whether there is a
+PriorityLevelConfiguration object with `Spec.Exempt==true`.  If not
+then the apiserver attempts to create the predefined configuration objects, the
+exempt PriorityLevelConfiguration last.  In this case: for each predefined
+object, the apiserver will create that object if there is not already
+an object with the same GroupKind and name; note that mere existence
+is enough; the apiserver does not insist that the spec matches its
+predefined content.
+
+Administrators can choose to persist a partially or completely
+different set of configuration API objects than the predefined set ---
+so long as the persisted set includes a PriorityLevelConfiguration
+object with `Spec.Exempt==true`.  (Remember that the behavior of the
+apiserver before it has had a chance to read the persisted
+configuration API objects is a different matter.)  Any configuration
+object can be created, updated, and deleted.
+
+#### API Server Initial Behavior
+
+As it starts up, an apiserver has to start handling requests before
+the filter discussed here has had a chance to read the persisted
+configuration API objects.  Indeed, that reading will have to be done
+through the very same filter (at least in the first apiserver in a
+cluster)!  The initial request handling also applies while an
+apiserver is persisting the default configuration API objects for the
+cluster.
+
+When the apiserver first starts up, it behaves as if the
+configuration consists of just (1) the predefined
+PriorityLevelConfiguration objects named `exempt` and `default`, and
+(2) the predefined FlowSchema objects named `exempt` and `default`.
+
+An apiserver maintains this initial behavior until it has had a
+chance to read all the persisted configuration API objects.
+
+After the initial behavior period is over, regular behavior begins.
+
+#### Default Request Handling
+
+It is possible that the administrators can modify the configuration
+API objects to create a situation where some requests will not match
+any FlowSchema.  This is also the situation before the cluster has
+been given its initial load of configuration API objects.  Once the
+initial behavior period is over, any request that does not match a
+FlowSchema object is handled as follows.
+
+If an unmatched request has `system:masters` among its user groups
+then the request is considered exempt and is assigned to one of the
+exempt priority levels.  If no PriorityLevelConfiguration object has
+`Spec.Exempt==true` then the apiserver acts as if there is an
+additional PriorityLevelConfiguration object with `Spec.Exempt==true`;
+for the rest of the `Spec` the apiserver will use either (a) what it
+most recently saw associated with `Exempt==true` or (b) the
+configuration as shown for the `exempt` PriorityLevelConfiguration
+above.
+
+An unmatched request that is not exempt is assigned to one of the
+priority levels with the "catchAll" bit set to true, with a flow
```

Comment:
We seem to have dropped this "catchAll" from the API?

Reply:
I think this paragraph needs an update and then we can merge this.

Reply:
The catch-all bit has been dropped from the current implementation but not the KEP.  This is a KEP update, so I felt free to refer to the catch-all bit.  Also, I have a PR against the feature branch to restore the catch-all bit (https://github.com/kubernetes/kubernetes/pull/82196), and will offer a revised version of that against `master` once we get our API object types into master.

Reply:
as a replacement of "catch-all" priority-level, we can have a flow-schema matching everything with logically lowest matching-precedance. what's why we dropped `catchAll`. the matching rules will be like:
```yaml
subject: [<authenticated>, <unauthenticated>]
apiGroups: [*]
resource: [*]
verbs: [*]
nonResrouceURLs: [*]
```

Reply:
If we have an non-deletable non-editable FlowSchema that matches everything then we do not need a catch-all bit.  I have been aiming to make every config object editable and deletable.